benchmarks/BenchmarkHelper.jl

```diff
@@ -0,0 +1,207 @@
+module BenchmarkHelper
```

Comment:
Hi @KDr2, a high-level comment here: can you take a look at the [`Nanosoldier.jl`](https://github.com/JuliaCI/Nanosoldier.jl) package, and see whether we can reuse it in Turing? For example, 

- Still use Github events to trigger action runners
- Use `Nanosoilders.jl` to parse benchmarking commands, and execute actual benchmarks (we probably need to rewrite our benchmarks using `BenchmarkTools.jl`, but it is ok to create a simple example for testing).
- Use `Nanosoilders.jl` to manage reporting, e.g. posting summary results to PRs, and archiving benchmarking results to a dedicated Github repo. 

This would better integrate with the Julia benchmarking ecosystem, and reduce the maintenance burden from our side in the long term.

Reply:
OK, I will add `BenchmarkTools` and `Nanosoilders` in to run benchmarks and parser command + generate reports.

But I think, to post benchmark reports to the main Turing repo maybe not a good idea since the reports could be increasingly large hence exploring the code base and messing up the commit history.  Or, if we want, we can commit the reports to the `gh-pages` branch, so that we can list and visit these reports on our website.

Reply:
> But I think, to post benchmark reports to the main Turing repo maybe not a good idea since the reports could be increasingly large hence exploring the code base and messing up the commit history. Or, if we want, we can commit the reports to the gh-pages branch, so that we can list and visit these reports on our website.

Let's keep these reports under a new repo, say `TuringLang/BenchmarkReports`.

Reply:
!benchmark(HEAD)

Reply:
!benchmark(HEAD)

Reply:
!benchmark(HEAD)

Reply:
!benchmark(HEAD)

Reply:
!benchmark(HEAD)

Reply:
!benchmark(HEAD)

Reply:
!benchmark(HEAD)

Reply:
!benchmark(HEAD)

Reply:
`runbenchmarks(ALL, vs="HEAD")`

Reply:
`runbenchmarks(ALL, vs="HEAD")`

Reply:
`runbenchmarks(ALL, vs="HEAD")`

Reply:
`runbenchmarks(ALL, vs="HEAD")`

Reply:
`runbenchmarks(ALL, vs="HEAD")`

Reply:
`runbenchmarks(ALL, vs="HEAD")`

Reply:
Update:
1. Now benchmarks are organized with `BenchmarkTools.jl` by adding new benchmarks to the root group `BenchmarkHelper.BenchmarkSuite`. All of them can be tagged.
2. The trigger command is Nanodoldier compatible: `runbenchmarks(TAGS, vs="BRANCH")`. In the command, a tag and a branch must be specified to filter and run the benchmarks.
3. The benchmark results are reported as comment reply of the triggering comment, in the format of DataFrame displaying, see https://github.com/TuringLang/Turing.jl/pull/1100#commitcomment-37342446.

I think, this PR now is ready to merge, as the first stage of benchmarking refactor. We can start another stage to integrate with `Nanosoldier`, Nanosoldier has much code of managing git repo and github event, which we may don't need.

@yebai

Reply:
Sure, sounds good to investigate Nanosoldier in a separate PR!